src/treesa.jl

```diff
@@ -156,82 +116,76 @@ Optimize the einsum contraction pattern using the simulated annealing on tensor
     end
+
     ###### Stage 2: computing ######
-    # create vectors to store optimized 1). expression tree, 2). time complexities, 3). space complexities, 4). read-write complexities and 5). slicing information.
-    trees, tcs, scs, rws, slicers = Vector{ExprTree}(undef, ntrials), zeros(ntrials), zeros(ntrials), zeros(ntrials), Vector{Slicer}(undef, ntrials)
+    # create vectors to store optimized 1). expression tree, 2). time complexities, 3). space complexities, 4). read-write complexities.
+    trees, tcs, scs, rws = Vector{ExprTree}(undef, ntrials), zeros(ntrials), zeros(ntrials), zeros(ntrials)
     @threads for t = 1:ntrials  # multi-threading on different trials, use `JULIA_NUM_THREADS=5 julia xxx.jl` for setting number of threads.
+
         # 1). random/greedy initialize a contraction tree.
         tree = _initializetree(code, size_dict, initializer; greedy_method=greedy_method)
-        # 2). optimize the `tree` and `slicer` in a inplace manner.
-        slicer = Slicer(log2_sizes, nslices, Int[labels[l] for l in fixed_slices])
-        optimize_tree_sa!(tree, log2_sizes, slicer; sc_target=sc_target, βs=βs, niters=niters, sc_weight=sc_weight, rw_weight=rw_weight)
+
+        # 2). optimize the `tree` in a inplace manner.
+        optimize_tree_sa!(tree, log2_sizes; sc_target=sc_target, βs=βs, niters=niters, sc_weight=sc_weight, rw_weight=rw_weight)
+
         # 3). evaluate time-space-readwrite complexities.
-        tc, sc, rw = tree_timespace_complexity(tree, slicer.log2_sizes)
+        tc, sc, rw = tree_timespace_complexity(tree, log2_sizes)
         @debug "trial $t, time complexity = $tc, space complexity = $sc, read-write complexity = $rw."
-        trees[t], tcs[t], scs[t], rws[t], slicers[t] = tree, tc, sc, rw, slicer
+        trees[t], tcs[t], scs[t], rws[t] = tree, tc, sc, rw
     end
+
     ###### Stage 3: postprocessing ######
     # compare and choose the best solution
-    best_tree, best_tc, best_sc, best_rw, best_slicer = first(trees), first(tcs), first(scs), first(rws), first(slicers)
+    best_tree, best_tc, best_sc, best_rw = first(trees), first(tcs), first(scs), first(rws)
     for i=2:ntrials
         if scs[i] < best_sc || (scs[i] == best_sc && exp2(tcs[i]) + rw_weight * exp2(rws[i]) < exp2(best_tc) + rw_weight * exp2(rws[i]))
-            best_tree, best_tc, best_sc, best_rw, best_slicer = trees[i], tcs[i], scs[i], rws[i], slicers[i]
+            best_tree, best_tc, best_sc, best_rw = trees[i], tcs[i], scs[i], rws[i]
         end
     end
     @debug "best space complexities = $best_tc, time complexity = $best_sc, read-write complexity $best_rw."
-    if best_sc > sc_target
-        @warn "target space complexity not found, got: $best_sc, with time complexity $best_tc, read-write complexity $best_rw."
-    end
-    # returns a sliced einsum we need to map the sliced dimensions back from integers to labels.
-    return SlicedEinsum(get_slices(best_slicer, inverse_map), NestedEinsum(best_tree, inverse_map))
+
```

Comment:
Do we need `sc_target` anymore? Maybe just remove it and add a new parameter `sc_weight`.